app/models/manageiq/providers/redhat/infra_manager/admin_ui.rb

```diff
@@ -0,0 +1,96 @@
+module ManageIQ::Providers::Redhat::InfraManager::AdminUI
+  extend ActiveSupport::Concern
+
+  require 'openssl'
+  require 'ovirtsdk4'
+  require 'uri'
+  require 'json'
+
+  def queue_generate_admin_ui_url
+    task_opts = {
+      :action => "Generate oVirt Admin UI URL for EMS"
+    }
+
+    queue_opts = {
+      :class_name  => self.class.name,
+      :instance_id => id,
+      :method_name => "generate_admin_ui_url",
+      :priority    => MiqQueue::HIGH_PRIORITY,
+      :role        => "ems_operations",
+      :zone        => my_zone,
+      :args        => []
+    }
+
+    MiqTask.generic_action_with_callback(task_opts, queue_opts)
+  end
+
+  def generate_admin_ui_url
+    $rhevm_log.info("Generating oVirt Admin UI URL for EMS with identifier '#{id}'.")
+
+    # Get the credentials of the provider:
+    user = authentication_userid(:default)
+    password = authentication_password(:default)
+
+    # Create the HTTP client:
+    insecure = default_endpoint.verify_ssl == OpenSSL::SSL::VERIFY_NONE,
+    ca_certs = default_endpoint.certificate_authority
+    ca_certs = nil if ca_certs.blank?
+    ca_certs = [ca_certs] if ca_certs
+    client = OvirtSDK4::HttpClient.new(
+      :insecure => insecure,
+      :ca_certs => ca_certs,
+      :log      => $rhevm_log
+    )
+
+    # Request new SSO token using the provider credentials:
+    request = OvirtSDK4::HttpRequest.new
+    request.method = :GET
+    request.url = URI::HTTPS.build(
+      :host  => default_endpoint.hostname,
+      :port  => default_endpoint.port,
+      :path  => "/ovirt-engine/sso/oauth/token",
+      :query => URI.encode_www_form(
+        :grant_type => "password",
+        :username   => user,
+        :password   => password,
+        :scope      => "ovirt-app-admin ovirt-app-portal"
+      )
+    ).to_s
+    request.headers = {
+      "Accept"     => "application/json",
+      "User-Agent" => "manageiq-providers-ovirt / API version #{api_version}"
+    }
+    client.send(request)
+    response = client.wait(request)
+
+    # Close the HTTP client:
+    client.close
+
+    # Extract the SSO token, return nil in case of error:
+    if response.code != 200
```

Comment:
The `response` can also be an exception, for example if there was a timeout when talking to the server. You may want to take that into account:

```ruby
if response.is_a?(OvirtSDK4::Error) || response.code != 200
  ...
end
```

Reply:
OK, I'll update the code accordingly.

I find this kind of convention (response is either HTTP response object, or an exception) a bit weird, I'd normally expect the relevant operation (like `send`) to throw an exception and the calling code could `rescue` that exception.

Reply:
The reason for this unusual interface is that the HTTP client is inherently asynchronous, and it processes multiple requests simultaneously. When you call the `send` method you are just queuing the request, for later execution. It is later, when you call `wait` that your requests (and other that may be queued) are really processed. The `wait` method returns when your request is processed, but many other request may have already been processed as well. It isn't right to throw an exception for those other requests which may have failed. Internally the HTTP client puts your request in a table, together with the result, which can be a response or an exception. When you call `wait` you are waiting for the result of a particular request, and then looking it up in that table. So it felt natural to just return that result, regardless of what it is. Could have been more natural to check if that result is an exception, and then raise it, but preferred to keep the interface closer to the internal implementation. This allows future improvements like waiting for multiple results.

Reply:
Thanks for the details!

> Could have been more natural to check if that result is an exception, and then raise it, but preferred to keep the interface closer to the internal implementation.

This is what I had in mind 😃 when calling `client.wait(request)` you're passing a specific `request` object, so I'd expect this operation to throw when that particular request fails. But keeping the interface closer to internal impl. means less code (and bugs) so I understand the motivation here.